src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java

```diff
@@ -1070,6 +1096,32 @@ private UsernamePasswordAuthenticationToken loginAndSetUserData(
             user.setFullName(fullName);
         }
 
+        // Set avatar if possible
+        if (StringUtils.isNotBlank(avatarFieldName)) {
+            try {
+
+                // Extract avatar from the userInfo
+                String avatarUrl = determineStringField(avatarFieldExpr, idToken, userInfo);
+                byte[] decodedAvatar = downloadAvatar(avatarUrl);
+
+                // Try to determine the content type of the avatar
+                String contentType = determineAvatarContentType(decodedAvatar);
+                LOGGER.finest("Avatar content type: " + contentType);
+
+                OicAvatarProperty.AvatarImage avatarImage = new OicAvatarProperty.AvatarImage(contentType);
+                OicAvatarProperty oicAvatarProperty = new OicAvatarProperty(avatarImage);
+                File targetFile = new File(user.getUserFolder(), "oic-avatar." + avatarImage.getFilenameSuffix());
+
+                Files.write(targetFile.toPath(), decodedAvatar, StandardOpenOption.CREATE, StandardOpenOption.WRITE);
```

Comment:
why would we have the complexity of downloading and storing the avatar and not just use its URL directly from `avatarProperty.getAvatarUrl()` ?

Reply:
> It's unlikely avatar images are stored on the same domain than Jenkins

Whilst I agree, I would consider it unlikely that arbitrary domains are used, rather one domain would be used for this and the admin can add this in the `img-src` CSP.

https://ci.jenkins.io/ via the GitHub branch source uses various 3rd party images (e.g. the stapler in https://ci.jenkins.io/job/Stapler/)

That said I will defer to someone with more knowledge of CSP than me, but if we need to download and cache avatar images (or others), then I would say that this should be something that core should handle for us asynchronously rather than having to rewrite this in multiple plugin and having the issues with potentially blocking login etc.

Reply:
Thanks for the input. 

I guess the option of direct link is much easier for now since it will not cause CSP issue for now.

And will remove the concern of login timeout.

Fine for you @jtnord ?